core/domain/acl_decorators_test.py

```diff
@@ -1024,3 +1026,53 @@ def test_exploration_editor_can_access_learner_dashboard(self):
         response = self.testapp.get('/mock/', expect_errors=True)
         self.assertEqual(response.status_int, 200)
         self.logout()
+
+
+class EditTopicDecoratorTest(test_utils.GenericTestBase):
+    """Tests the decorator can_edit_topic."""
+    username1 = 'topicmanager'
+    user_email1 = 'user@example.com'
+    username2 = 'normaluser'
+    user_email2 = 'user2@example.com'
+    topic_id = 'topic_1'
+
+    class MockHandler(base.BaseHandler):
+        @acl_decorators.can_edit_topic
+        def get(self, topic_id):
+            self.render_json({'topic_id': topic_id})
+
+    def setUp(self):
+        super(EditTopicDecoratorTest, self).setUp()
+        self.signup(self.ADMIN_EMAIL, self.ADMIN_USERNAME)
+        self.signup(self.user_email1, self.username1)
+        self.signup(self.user_email2, self.username2)
+        self.set_admins([self.ADMIN_USERNAME])
+        self.set_topic_managers([self.username1])
+
+        self.admin_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+        self.manager_id = self.get_user_id_from_email(self.user_email1)
+        self.admin = user_services.UserActionsInfo(self.admin_id)
+        self.manager = user_services.UserActionsInfo(self.manager_id)
+
+        self.testapp = webtest.TestApp(webapp2.WSGIApplication(
+            [webapp2.Route('/mock/<topic_id>', self.MockHandler)],
+            debug=feconf.DEBUG,
+        ))
+        topic_services.create_new_topic_rights(self.topic_id, self.admin_id)
+        topic_services.assign_role(
+            self.admin, self.manager_id, topic_domain.ROLE_MANAGER,
+            self.topic_id)
+
+    def test_topic_manager_can_edit_topic(self):
+        self.login(self.user_email1)
+        response = self.testapp.get(
```

Comment:
Should be using self.get_json, not self.testapp.get()?

Reply:
I am not sure. Here we are testing on the mock handler. get_json uses all routes from main.py which include all the actual controllers. To send the request to the mock handler only, we use a separate testapp.

Reply:
You could perhaps do something like this:

```
mock_testapp = ...

with self.swap(self, 'testapp', mock_testapp):
    ...
```

Reply:
I'll try this out. Thanks!

Reply:
Done